Implement Mmarkdown ☕️

albinotonnina opened the pull request:
Hello Michele!
I wanted to try [mmarkdown](https://github.com/albinotonnina/mmarkdown) for the first time here, since it's this project that gave me the idea.

You can see my version of the new Readme [here](https://github.com/albinotonnina/css-in-js/tree/mmarkdown)

### What I did:
* mmarkdown! 😄 [code here](https://github.com/albinotonnina/css-in-js/tree/mmarkdown/ReadmeSrc). I call a githup API and `make` the markdown.

* moved files: Since the Readme got longer, and since I linked the demo folder under each package, I thought that removing some scrolling for the users by moving the demo packages in a subfolder was a good idea :)

* added a link to mmarkdown in the readme 😅

### What I'm adding
Stars, last npm version, repo description, open issues, forks, author data, license, creation date, archived state.
Each package in the table links to a package 'card' down in the file

### What is not that right
I created a data.json file with the features you produced. Realized only later that you already have one and use it for both the web app (which again I noticed too late) and the readme file.
What I propose to solve this is to generate a json file from my new code and have that for the webapp.

Also, I used a trick to save on ajax calls to the github APIs.
I created a new github user and I just starred every repo in the project. Then I call a user's `/starred` API to get the list and the data. This is a bit smelly.

streamich commented:
Can you sort the packages alphabetically as in the current readme?

albinotonnina commented:
Hi @streamich,
They are ordered alphabetically as a list, after the table.
I thought that listing them by star would be an important factor for people to decide what library to use in the end. You disagree?

albinotonnina commented:
Or we could derive a maybe more fair evaluation system from the stars, such as:
'very popular'
'popular'
'less popular'

something along those lines.

streamich commented:
I was looking for packages and realized they are not alphabetically ordered, so it was hard to find them. Besides that here is what problems I see when ordering by stars:

1. The whole purpose of this repo is to let people to compare packages by their features, not stars. This way I feel that it is a bit counter productive if people just choose the most starred packages.
2. It is also self-reinforcing: `styled-components` is already the most popular package, and if more people pick it just because it has more stars, its lead is just going to increase.
3. Stars are misleading:
  1. `emotion` is basically a clone of `styled-components` but superior in every respect, however, it has 5x less stars than `styled-components`.
  2. `radium` has a lot of stars because it is one of the first libraries, but it is one of the worst options nowadays, as it uses inline styles, and most of the other libraries actually emit CSS.
4. But what you want to instead is for developers to see new good libraries.

I would rather order them alphabetically, or if you want to have a fancy sorting algorithm, it could include the following factors:

1. Newer libraries should be pushed towards the top, just like on Reddit or Hacker News.
2. Number of stars should positively influence ranking.
3. Possibly other factors could influence the score as well:
  1. *negative*: Open issues
  2. *positive*: Closed issues
  3. *positive*: Number of releases on NPM

Here is Hacker Rank formula that can be used:

```js
const score() => (stars, unixTimeStampWhenCreated) => {
    const hoursSinceCreated = (Date.now() - unixTimeStampWhenCreated) / (1000 * 60 * 60);
    return stars / Math.pow(hoursSinceCreated + 2, 1.5);
};
```

BTW, `mmardown` is a nice project, I am planning to use it somewhere as well.

albinotonnina commented:
Agreed, you have more than one point.
I toned down the stars part and I ordered the packages by name.
The API I'm using is missing important data like closed issues, watchers, releases etc.
This would be more interesting if someone would spend some getting the fancy sorting algorithm. 
We will need multiple calls on Github and some work. 
Very interesting though.

MicheleBertoli commented:
Thank you very much @albinotonnina for your PR, and thank you very much @streamich for your comments (I 100% agree with everything you said).

`mmarkdown` is cool, and I like the idea of fetching more information about the packages.
However, the resulting [README](https://github.com/albinotonnina/css-in-js/blob/05ad9b353e587cb6477bb2a7e34fe54de214843b/README.md) is too long and hard to read.

Random thoughts:
- I would avoid listing the single packages and stick to the table.
- the code examples are relevant and they should be in root folder.
- I would "merge" some of the new columns (e.g. open issues, stars, created) and expand "features".
- "last updated" seems more interesting than "created".

What do you think?

albinotonnina commented:
Thank you both for your feedback! Very appreciated!
[New version here](https://github.com/albinotonnina/css-in-js/tree/mmarkdown)
Let me know (tell me if it somehow it look funny)

This version solves the problems of the previous one.
New caveats: you need a a [github access token](https://github.com/settings/tokens) and set an env variable `process.env.GITHUB_ACCESS_TOKEN`
Also, now the fetching takes around 5 minutes. I have to pause every 5 seconds between packages otherwise Github complains.

streamich commented:
I think GitHub token env variable is usually called `GITHUB_TOKEN`.

albinotonnina commented:
And we like conventions :) I changed that.

albinotonnina commented:
There something I missed, the archived attribute from github. I think a couple of them were set as archived. Some of them had just the deprecated message in the description field still in github.

stereobooster commented on .gitignore:
```diff
@@ -2,4 +2,10 @@ lib/
+.vscode/launch.json
+.vscode/settings.json
+.env
+.prettierrc
```
this one should not be ignored. If you use some settings for prettier, then commit it. Or use default settings for prettier

albinotonnina replied:
I like no-semi, my .prettierrc has no semi. What should I do, commit?

albinotonnina commented on .gitignore:
```diff
@@ -2,4 +2,10 @@ lib/
+.vscode/settings.json
+.env
+.prettierrc
+yarn.lock
```
Thank you. I do it for my repo. Never sure what to do with others.

stereobooster commented on ReadmeSrc/Readme.js:
```diff
@@ -0,0 +1,142 @@
+  });
+}
+
+async function getNpmDownloads(npm) {
```
you can write this shorter

```js
const getNpmDownloads = (npm) => axios
  .get(`https://api.npmjs.org/downloads/point/last-week/${npm}`)
  .then(response => ({ downloads: response.data.downloads }))
```

albinotonnina commented:
Pushed changes @stereobooster  suggests

stereobooster commented on ReadmeSrc/Rows.js:
```diff
@@ -0,0 +1,65 @@
+      const spacer = index > 0 ? '<div>____________</div>': ''
+
+      //prettier-ignore
+      const openIssues =`<div><sub><img src="./ReadmeSrc/img/hand-paper-o.svg" width="12" />  ${stat.githubData.open_issues}</sub></div>`
```
```js
const divWithIcon = (icon, value) => 
  `<div><sub><img src="./ReadmeSrc/img/${icon}.svg" width="12" />${value}</sub></div>`
const openIssues = divWithIcon("hand-paper-o", stat.githubData.open_issues);
```

Also not sure why you want to use `//prettier-ignore`. Formatting here does not contain any value. Idea of prettier is to stop worrying about formatting. And you need `//prettier-ignore` only when, formatting itself is valuable.

albinotonnina replied:
I didn't like the way prettier formatted that part but ok, I'll remove the ignore comments.

![image](https://user-images.githubusercontent.com/37283/39299312-56c6bb7c-4940-11e8-8633-1a4ab360b4fc.png)

stereobooster replied:
> I didn't like

This is the idea. You just need to stop care. I do not like semicolons either, but I'm using prettier without config (e.g. with semicolons). And I stopped to care what is happening with formatting, because I do not need to touch it. It is not my job. This is the part of the code which I will never need to touch.

albinotonnina replied:
You are right. 
I removed them

stereobooster replied:
```js
const divWithIcon = (icon, value) => 
  `<div><sub><img src="./ReadmeSrc/img/${icon}.svg" width="12" />${value}</sub></div>`
const openIssues = divWithIcon("hand-paper-o", stat.githubData.open_issues);
```

stereobooster commented on ReadmeSrc/Readme.js:
```diff
@@ -0,0 +1,116 @@
+
+const {getRows, headers} = require('./Rows')
+
+const data = require('../webpage/src/data.json')
```
It would make sense to save fetched data back to `/webpage/src/data.json` this way it can be reused by webpage.

albinotonnina replied:
Yes. I see this as a separate PR though. I would do something more with the new data and the web app, sorting, filtering, more data.

stereobooster replied:
Yes I do not ask you to change web application. I'm asking to write this data to JSON file.

```js
const writeOut = (data, path) =>
  new Promise((resolve, reject) =>
    fs.writeFile(path, JSON.stringify(data, null, 2), "utf-8", err => {
      if (err) reject(err);
      else resolve(data);
    })
  );
await writeOut(data, '../webpage/src/data.json')
```

stereobooster commented on ReadmeSrc/Readme.js:
```diff
@@ -0,0 +1,116 @@
+
+const data = require('../webpage/src/data.json')
+
+let rows
```
I understand this is one-time-run script and this is not critical, but i prefer to avoid global state. It is hard to read this code. You can encapsulate `let rows` inside function which is using it and latter returns it.

albinotonnina replied:
That's not global state. 
https://github.com/albinotonnina/mmarkdown/blob/master/src/extract.js#L51

stereobooster replied:
This is global state for functions inside of this file. E.g. order function calls matters, calling same function twice can change result of function

albinotonnina replied:
Anyway yes. Of course I had the same thought of your with that. 
And I agree with you, it's not critical.
But I changed that.

stereobooster commented:
@albinotonnina thank you for your contribution. I didn't mean to overload you with those requests.

My general idea - we can separate script which fetches data and script which generates readme. We can ask @MicheleBertoli opinion on this.

albinotonnina commented:
No I appreciated the requests. Some of them I missed some of them I would have done them with another pass on the code. The code separation also made sense to me since the beginning.
I wish I had more time today to do it but unfortunately I really cannot.
Thanks for your review really, appreciated it.

albinotonnina commented:
Ok I couldn't go without doing at least something about it.
@stereobooster now the code is better separated.

stereobooster commented:
Let me explain my motivation. I believe that `Readme.md` and `data.json` should be in sync, so you can always regenerate the Readme from data.json and result will be the same. If we will keep up-to-date `data.json` it can be reused by other projects as the source of truth. 

> Also I'm still not sure about overriding data.json with the fetched data.
People contributing to the repo will think they have to fill those values.

Good point - we can create section called `autogenerated`, so people would understand they do not need to touch it, for example:

```json
"Styles As Object Literals": true,
"Extract CSS File": true,
"autogenerated": {
  "downloads": 100000,
  ...
}
```

We will have two separate scripts:
1. update stats - download data from github and npm and populate to `data.json`
2. generate readme - generate readme based on `data.json`

---

Other note about `data.json` format. You can change this

```json
     {
       "Package": [
         {
           "text": "aphrodite",
           "href": "https://github.com/Khan/aphrodite"
         }
       ],
      "Package Stats": [
        {
          "npm": "aphrodite",
          "github": "Khan/aphrodite"
        }
      ],
```

into this:

```json
     {
       "Package": [
         {
           "text": "aphrodite",
           "href": "https://github.com/Khan/aphrodite",
           "npm": "aphrodite",
           "github": "Khan/aphrodite"
         }
       ],
```

Data is the same, but a bit more compact format

stereobooster commented on ReadmeSrc/Rows.js:
```diff
@@ -0,0 +1,69 @@
+const headers = [
+  'Package',
+  //   'Version',
```
this is the version of the package used in example (or isn't it?). I would still show it. And also show latest version in npm - to be able to detect how given example is outdated.

stereobooster commented on ReadmeSrc/Rows.js:
```diff
@@ -0,0 +1,69 @@
+    .map((stat, index) => {
+      const spacer = index > 0 ? '<div>____________</div>' : ''
+
+      const openIssues = `<div><sub><img src="./ReadmeSrc/img/hand-paper-o.svg" width="12" />  ${
```
```js
const divWithIcon = (icon, value) => 
  `<div><sub><img src="./ReadmeSrc/img/${icon}.svg" width="12" />${value}</sub></div>`;

const openIssues = divWithIcon("hand-paper-o", stat.githubData.open_issues);
const closedIssues = divWithIcon("hand-peace-o", stat.githubData.closed_issues);
```

albinotonnina commented:
Hello @stereobooster,
Thank you again for your comments!
All fair and all done.

Moved the fetch data scripts in /webpage amongst the rest.

stereobooster commented on ReadmeSrc/Readme.md:
```diff
@@ -0,0 +1,60 @@
+
+If your package is not listed here, feel free to add it.
+
+1.  Create a new folder named `package-name` in `Packages`.
```
in the root folder

albinotonnina replied:
Changed.

stereobooster commented on ReadmeSrc/Readme.md:
```diff
@@ -0,0 +1,60 @@
+1.  Create a new folder named `package-name` in `Packages`.
+2.  Implement the red button example using the package.
+3.  Add a new entry to [data.json](webpage/src/data.json).
+4.  Re-generate the data with: `cd webpage && yarn && yarn update-data`.
```
`data.json` is located in `webpage/src`, because Create-React-App does not allow require files outside of `src`, otherwise I would move it to `ReadmeSrc`. We can use symlink to move file, but this will not work on windows.

People do not need to install all dependencies of webpage to manipulate `Readme`, as soon as we have separate project responsible for readme. So I would move fetch-data script to ReadmeSrc.

```
cd  ReadmeSrc && yarn && yarn update-data && yarn generate-readme
```

albinotonnina replied:
I just did it. There is one issue, users that launch generate-data need a github access token.
I added a `.env_example` file and a bullet point in the instructions. what do you think?

MicheleBertoli commented:
Thank you very much @albinotonnina @stereobooster.
I'm sorry if I didn't participate in this discussion in the last few days, but I was on vacation.
I'll check this out ASAP.

albinotonnina commented:
I imagined @MicheleBertoli, hope you had fun :)
I just pushed a new commit with updated data.

MicheleBertoli commented:
Thank you very much one more time @albinotonnina @streamich for your hard work on this.

I checked the [latest version](https://github.com/albinotonnina/css-in-js/blob/5ef4d63941f6c3d8e6170d3305e4ec2fce0d0145/README.md) of the README, and I have a few comments/questions:

- the rows are still too high, I wouldn't add more than one line. What about something like this?

<img width="890" alt="screen shot 2018-05-12 at 6 59 05 pm" src="https://user-images.githubusercontent.com/1308971/39960233-fd77963e-5616-11e8-91b9-6ebd5bf86d5f.png">

- are the SVG icons free?

- apart from the "issue" icon (which I would change with the exclamation mark, see GitHub) the other icons are pretty understandable and I would remove the legend (nice to have: a tooltip on hover).

- I would put the "last updated" sentence below the table.

- I don't think the changes to the `.gitignore` and `.prettierrc` are needed.

- I would prefer to have a single command (e.g. merging `update-data` and `generate-readme`) and a different way to pass the GitHub token (e.g. ENV variable). I also think that asking people to generate an access token is a bit too much, can we move this logic in the CI?

albinotonnina commented:
Hello @MicheleBertoli, thanks for the feedback.

- Gotcha. Check the [latest version](https://github.com/albinotonnina/css-in-js/tree/mmarkdown#_)

- Yes, Icomoon free version: [GPL / CC BY 4.0](https://icomoon.io/#icons-icomoon)

- Changed the icon, added tooltip.

- Would you delete `.prettierrc`? I could revert `.gitignore` but then could you suggest me how to ignore `.prettierrc` and my `.vscode files` without changing `.gitignore`?

- Would you add a new package.json in the root with the one merged command? The token is already passed with an [ENV variable](https://github.com/albinotonnina/css-in-js/blob/mmarkdown/ReadmeSrc/.env_example). What would you suggest for the CI setup?

MicheleBertoli commented:
WHOA, thank you @albinotonnina for applying all the changes - I love the way it looks now.

I'm not entirely sure what's the best way to avoid committing your personal prettier and vscode files.
I use [Nuclide](https://nuclide.io) and I don't have those.

I'd be nice to have one single command on the root (e.g. `yarn build`) as it would make it easier for people to contribute.
For the ENV variable, sorry if I wasn't clear, I meant to say we could tell people to pass it within the command rather than editing a file, like: `GITHUB_TOKEN=123abc yarn build` (but I can also see the benefits of storing it).

Finally, I'm not a CI expert but I would love to automate the build process of the README (e.g. on every commit to master). Any idea @stereobooster?

streamich commented:
Would be nice if "npm version" icon was a link, when user clicks that, it would go to NPM page.

albinotonnina commented:
Hi @streamich, 
Thanks, it makes sense. Is that ok for you @MicheleBertoli?

MicheleBertoli commented:
+100 on the links, they would make the table even more useful.
Thank you very much!

stereobooster commented:
> Finally, I'm not a CI expert but I would love to automate the build process of the README (e.g. on every commit to master). Any idea @stereobooster?

You can push from Travis back to github - to do that you check that last commit is not from travis (to prevent loop), build Readme and push back to branch. We already pushing from Travis to github (gh-pages) - it is ok because bot pushes to one dedicated branch, if it will push to the same branch as developer it would be confusing from my POV, but maybe everybody else will be ok with it.

Other idea, maybe we try push to much functionality into readme, maybe it is time to refuse from old approach, because it doesn't scale and add all new feature to website instead of readme. Just an idea.

albinotonnina commented:
@MicheleBertoli @streamich: links added

stereobooster commented on ReadmeSrc/scripts/fetchData.js:
```diff
@@ -0,0 +1,89 @@
+const data = require('../../webpage/src/data.json')
+
+// https://github.com/settings/tokens
+const access_token = process.env.GITHUB_TOKEN
```
@MicheleBertoli 

you can do this

> `GITHUB_TOKEN=123abc yarn build`

MicheleBertoli replied:
Oh yes, I was more referring about what we tell people to do in the new README:
> 4. Rename ReadmeSrc/.env_example to ReadmeSrc/.env and set a Github access token to retrieve data from Github.

streamich commented:
@stereobooster In your commit message you add `[skip ci]` so that Travis does not build that commit.

MicheleBertoli commented:
I agree @streamich I also think we should merge this first, and then we can build some infra.

albinotonnina commented:
@MicheleBertoli @streamich I would go for the separate PR.

albinotonnina commented:
@streamich A👏MA👏ZING👏
Hope it’s doing well there!

MicheleBertoli commented:
@albinotonnina do you mind resolving the conflicts so that we can merge?
Thank you very much!

albinotonnina commented:
Morning @MicheleBertoli!
Conflict resolved

MicheleBertoli commented:
🙌
